### ISSUE TITLE
Forced subtitles

### ISSUE THREAD

**Issue by Varstahl** [Sandmann79/xbmc#113]
After @muphus suggestion I merged his enhancement into a separate [feature-forcedsubs](https://github.com/Sandmann79/xbmc/tree/feature-forcedsubs) branch, and applied a few fixes. Before merging it into master we need to decide/figure out what to do with the subtitles (again).

We previously completely disregarded user's choice in the subtitles settings when it came to non-languages. Subtitles can return, on top of languages, four values: `none`, `forced_only`, `original` and `default`.

With the [latest commit](https://github.com/Sandmann79/xbmc/commit/a1eb042a8f858718ef59686d97dddc2ba7de561c) I've treated:
* **none**: as imperative, discarding all subs found;
* **forced_only**: as itself, pulling the language from the user's audio track preference (possibly wrong, more on that later);
* **original**: while I believe amazon actually returns the default stream language (at least for audio, don't remember if it does for subtitles too) it required some more reverse engineering I didn't have the time to do. For the time being I've opted to treat it as `default`;
* **default**: labeled as `Kodi UI language`, it just says to use whatever language kodi's UI is in.

The `forced_only`/`original` settings cause further branching since even audio language can specify both `original` and `default` non languages. I've treated `original` audio language as `default` for the subtitles part.

In regard to the forced subtitles it might be possible/welcome to download all types of forced, since in theory it should change along with the language. Although Kodi kind of butchers this feature (didn't re-test after last nightly build update though, it might be fixed by now).

So, now, we are left to decide what we want to do with the settings. My opinion would be to split the current subtitle and language into two parts, one for the subtitles management (Kodi settings / Override kodi settings with audio language / Override kodi settings with forced only / whatever) and one for the fallbacks (If subtitle language is not available fallback to: none/audio language/en/all). These are just ideas, there's not really a right or wrong, but it seems to be the most sensible thing.

What do you guys think? @Sandmann79, @muphus, opinions?

**Comment by Varstahl** [comment]
> BTW I've just discovered that if you replace the subs' url "extension" from .ttml2 to .srt, you will get a plain srt file from PrimeVideo servers, so no conversion needed anymore. Moreover srt files are ~50% smaller than ttml2 files, so lower download times.

Hah, figures. I mostly didn't touch/reverse engineer the process past the introduction of PrimeVideo, but if what you're saying is true and works on Amazon.tlds as well, it could pretty much make everything faster by quite a bit.

> IIRC all this hassle of choosing which subtitles to download is needed to avoid very long waiting times due to file conversions

Yes, that took a part in it. On the other hand it probably makes the most sense to give the users the best possible options. I, for instance, might only need italian subtitles with english fallback. Although I could also read different languages, I probably won't ever need arabic, hindi and chinese. Furthermore several people have still issues with slow connection speed or monthly bandwidth cap. So it makes no sense to download useless stuff every time a video is played, if we can avoid it.

Hell, to make it better we could even add an additional caching storage file and avoid downloading subtitles multiple times for the same video based on its ASIN, and maybe add a TTL to it. @Sandmann79 sounds like a good idea? :)

> I really don't know how this situation should be handled

There is no real right or wrong here, or to put it in marketing terms "there is no perfect pasta sauce, only perfect pasta sauces". This could be handled in several different ways, it's just a matter of collectively deciding which path to take while keeping the overall complexity relatively low. A good discussion is all we need :)

**Comment by muphus** [comment]
> Furthermore several people have still issues with slow connection speed or monthly bandwidth cap. So it makes no sense to download useless stuff every time a video is played, if we can avoid it.

Yeah, absolutely true. I've quickly tried to bypass the conversion by using `*.srt` urls but the final result is not so different from the current performances. At first I thought that Kodi would download `http://` subs "on-demand" when selected by the user, but it seems that it downloads every sub before the video playback. Moreover if we directly pass `http://` urls to Kodi as subs, it won't recognize their language because of the non-standard filenames (e.g.: `f75d9811-e86b-40ed-a47a-5021a33abedb.srt`), so we need to download them manually to set the correct filename as we already do now.

So yes, a language selection is really needed.

**Comment by Varstahl** [comment]
That's mostly a problem with the lack of parallelization over multiple files. It could be possible to reduce by starting several threads instead of  using a single connection to get all 20 or so of them. But the problem still stands, even though possible it makes no sense to do so for most people, so even spending time on it sounds silly.

**Comment by Varstahl** [comment]
Since @Sandmann79 has gone silent for more than a month, I'll go ahead and complete everything myself even though I can't test on amazon.tld accounts, and pray for the best.

I've decided to go with this:
* None
* All
* From Kodi player language settings
* From settings, fallback to english
* From settings, fallback to all

In addition to that I'm introducing a new toggle option, that lets the user select between regular subtitles and forced subtitles. This way we cover all bases without making the subtitle choice setting a mess.

I just need to implement the few changes in the code we need, then I'll deploy and merge.

**Comment by Varstahl** [comment]
Applied all the changes and as of 817b43d08916c9b58556447dcf3ebd321bb3d357 the episode load speed improved dramatically, thanks to forced narrative and `.srt`.

> Moreover if we directly pass http:// urls to Kodi as subs, it won't recognize their language because of the non-standard filenames (e.g.: f75d9811-e86b-40ed-a47a-5021a33abedb.srt), so we need to download them manually to set the correct filename as we already do now.

I gave this a thought today, and I realized that Kodi addons are just akin to web servers. In theory it should be possible to pass a special URL to Kodi instead of the direct subtitle stream name, something like `'{0}?mode=SubtitleStream&remote={1}&displayName={2}'.format(sys.argv[0], <base64encoded stream url>, <our language code format>)`. This could (very) theoretically work by proxying the file content, but I honestly can't be arsed to even try it, I'm just reporting the possibility for science :D
